Updates to Electride Finder and Bug Fixes

SWeav02 opened the pull request:
This PR primarily aims to fix an issue with the automatic electride placement based on the ELF's shape. The issue stemmed from the use of Pyrho to regrid the ELF which evidently resulted in slight changes in the shape of the ELF and slight shifts off of symmetrically equivalent sites for electrides.

The PR also addresses a known bug in the voronoi like plane partitioning where some voxels were not able to be assigned to a voronoi region. Holes in the voronoi surface turn out to be a known feature of weighted voronoi systems similar to the one we employ. A backup voxel assignment is added which assigns remaining voxels to the closest partitioning plane and therefore the atom which the partitioning plane belongs to.

The final update is to the electride dimensionality algorithm which has been improved to give all possible dimensionalities as the dimensionality can change if you use varying ELF values as a cutoff.

SWeav02 committed:
Add matrix property to grid

SWeav02 committed:
Add matrix property to grid

SWeav02 committed:
Update partitioning scheme

SWeav02 committed:
Update partitioning scheme

SWeav02 committed:
Update Voxel Assignment Scheme

SWeav02 committed:
Update Voxel Assignment Scheme

SWeav02 committed:
Fix bug in partitioning

SWeav02 committed:
Fix bug in partitioning

SWeav02 committed:
Small updates to voxel assignment

SWeav02 committed:
Small updates to voxel assignment

SWeav02 committed:
update indexing in partitioning

SWeav02 committed:
update indexing in partitioning

SWeav02 committed:
Add partitioning refinement

SWeav02 committed:
Add partitioning refinement

SWeav02 committed:
Fix bug with calculating voxel frac coords

SWeav02 committed:
Fix bug with calculating voxel frac coords

SWeav02 committed:
Small documentation improvement

SWeav02 committed:
Small documentation improvement

SWeav02 committed:
Small documentation improvement

SWeav02 committed:
Small documentation improvement

SWeav02 committed:
Small bug fix

SWeav02 committed:
Small bug fix

SWeav02 committed:
Fix major error in voxel assignment

For voxels split by planes, sites are found for each vertex of the voxel. Previously, sites were assigned even if the vertex sat on a plane. This caused some vertices to be assigned to multiple sites and caused an error.

SWeav02 committed:
Fix major error in voxel assignment

For voxels split by planes, sites are found for each vertex of the voxel. Previously, sites were assigned even if the vertex sat on a plane. This caused some vertices to be assigned to multiple sites and caused an error.

SWeav02 committed:
black and isort

SWeav02 committed:
black and isort

SWeav02 committed:
Update Voxel Assignment Doc Strings

SWeav02 committed:
Update Voxel Assignment Doc Strings

SWeav02 committed:
Update remaining Doc strings

SWeav02 committed:
Update remaining Doc strings

SWeav02 committed:
Remove plane reduction method

SWeav02 committed:
Remove plane reduction method

SWeav02 committed:
Fix plane selection bug in voxel assignment

SWeav02 committed:
Fix plane selection bug in voxel assignment

SWeav02 committed:
Fix bug in electride voxel assignment

SWeav02 committed:
Fix bug in electride voxel assignment

SWeav02 committed:
Merge pull request #2 from SWeav02/simple_plane_split

Remove plane splitting assignment

SWeav02 committed:
Add backup for when voxels are assigned to multiple sites

SWeav02 committed:
Add backup for when voxels are assigned to multiple sites

SWeav02 committed:
Fix bug in electride finder

SWeav02 committed:
Fix bug in electride finder

SWeav02 committed:
Remove Core parameter

With the new numpy based badelf, I haven't implemented a core selection. This only effects the Dask arrays that are used for larger voxel grids. I can probably add the parameter back in if it appears to be important.

SWeav02 committed:
Remove Core parameter

With the new numpy based badelf, I haven't implemented a core selection. This only effects the Dask arrays that are used for larger voxel grids. I can probably add the parameter back in if it appears to be important.

SWeav02 committed:
Merge remote-tracking branch 'origin/vectorize_badelf' into vectorize_badelf

SWeav02 committed:
supress warnings in context

SWeav02 committed:
Add BadelfConfig line when registering badelf

SWeav02 committed:
Remove cores parameter documentation

SWeav02 committed:
supress warning in context

SWeav02 committed:
Merge branch 'main' into vectorize_badelf

SWeav02 committed:
Add electride position refinement

SWeav02 committed:
Update Partitioning

There was a bug in the partitioning when two atoms of the same type were next to each other but not symmetric. This mostly effects electrides with the Voronelf alg.

SWeav02 committed:
Update electride dimensionality search

SWeav02 committed:
Update voxel coords to 0 base

SWeav02 committed:
Add more coordinate conversion methods

SWeav02 committed:
remove regrid and refinement from electride finder

Electrides were being placed at odd spots. Regridding seemed to contribute to this and the method I had for refining didn't seem to help. Its possible that I'll need to put a refinement method back in later, but that will depend on if it seems to effect anything in structures I havent tested yet.

SWeav02 committed:
Large Voxel Assignment Update

SWeav02 committed:
Add rounding to badelf results

SWeav02 committed:
Add electride structure and dimensionality to database

jacksund commented on src/simmate/apps/badelf/models/badelf.py:
```diff
@@ -15,6 +15,13 @@ class BadElf(Structure, Calculation):
     class Meta:
         app_label = "workflows"
```
I missed this in your last PR but this `Meta` tag and the app label need to be removed until badelf becomes a default app. Also the migrations in this MR need to be deleted and recreated (and they'll then show up in the badelf app).

All this will do is change your table name from `workflows_badelf` to `badelf_badelf` in the postgres database -- but organizing the workflow migrations is the bigger deal here and what would cause trouble for users + db maintainers.

SWeav02 replied:
Ok cool. I'm making a clean simmate dev install and remaking the migrations.

jacksund replied:
if you want the table to still show as `workflows_badelf` in the postgres db: https://docs.djangoproject.com/en/5.0/ref/models/options/#table-names

``` python
class Example(models.Model):
    . . .
    class Meta:
        db_table = "workflows_badelf"
```

SWeav02 committed:
Merge branch 'main' into update_electride_search

SWeav02 committed:
Remove meta and update migrations

SWeav02 commented:
@jacksund this PR is basically ready, but I'm going to be testing the new version of BadELF on ~50 electrides so I think we wait to merge until after these are done. That way I can fix any bugs that I find.

jacksund commented on src/simmate/website/workflows/migrations/0001_initial.py:
can you reset this file back to it's original? I'm not sure what it's changing, but it should be untouched. Maybe just copy/paste the content from the `main` branch into your branch's file here

jacksund replied:
Thanks 😄

If you need it: https://stackoverflow.com/questions/43275419/reverting-one-file-in-gitkraken
``` bash
# reset a file in git
git checkout main -- path/to/your/file
```

jacksund commented:
> I'm going to be testing the new version of BadELF on ~50 electrides

sounds good. If you have a script that's automating these tests, feel free to upload them. You can then run the checks whenever and use `math.isclose()` or a similar numpy function to make sure your getting expect answers for each electride

SWeav02 commented:
The tests aren't automated yet since I'm running BadELF on preran DFT and I'm not entirely sure what I'm expecting the results to be yet. I'm planning to look at them manually to make sure everything is going smoothly. I'll make a more automated version later though that I can upload in a future PR.

SWeav02 committed:
Fix migrations

SWeav02 committed:
Fix small bug in dimensionality alg

When 0 connections were found between electride sites, the dimensionality alg would throw an error. This has been corrected.

SWeav02 committed:
Tweeks to BadELF

This commit makes changes to fix a number of issues found when testing BadELF on ~50 electrides. The main change is the removal of one of the partitioning plane reduction algorithms which would sometimes remove important planes. An alternative version will need to be added in the future to reduce computation times. Other changes include the addition of a minimum electride radius in the electride_finder module and two new columns in the badelf Model.

SWeav02 committed:
Switch to Grid based Dimensionality

The previous dimensionality algorithm relied on connections between electride sites and the pymatgen get_dimensionality_larsen method. There were several issues where this method failed, some of which I haven't figured out. One fundamental issue was that the ELF can connect at a given isosurface level in ways that are not along the bonds between electride sites. The new method allows for these connections and is faster, but has the slight disadvantage that the connecting values are not interpolated and are therefore dependent on the voxel grid density. However, this is probably less than the error inherent to the ELF.

SWeav02 committed:
Add warning for printing ELFCAR/CHGCAR

There is currently no method for printing out voxels assigned using the zero-flux method. Added a simple warning message if users attempt to.

SWeav02 committed:
Fixed small bug in dimensionality alg

The dimensionality alg returned None instead of 0 when no grid features were found. This caused failures down the line.

SWeav02 committed:
Update get_elf_radius and covalency check

The _get_elf_ionic_radius method is primarily used in the electride finder. In some electrides there are sodium atoms surrounded by ELF which can be very far from atoms. There was already a distance check for these, but for one electride (Ba3na4N) this didn't work. I've added a check for if the bond is covalent.  I'm also testing a new method for checking structures for covalency and I've updated the covalency check accordingly.

SWeav02 committed:
Move all_voxel_frac_coords to Grid class

SWeav02 committed:
Update grid to use VolumetricData

SWeav02 committed:
Add to_pybader to Grid class

SWeav02 committed:
Update electride finder to basin based

SWeav02 committed:
switch to pybader

SWeav02 committed:
Add assignment recognition to partitioning

With pybader being incorporated I can now keep track of where the zero-flux elf basins are located when determining partitioning. I've changed how the partitioning frac is found from being the closest min to the center. It now finds the first point along the line where the assignment switch from the atom to the neighbor. If there are assignments along the line other than the two sites, it assigns the frac to the maximum closest to the center, as this is likely closer to the center.

SWeav02 committed:
Test fix for partitioning bug

With the updated partitioning scheme, there are some differences in where partitioning planes are located. This is most apparent in VoronELF. The main issue here was that I assigned the partitioning fraction based on the closest maximum to the center if voxels assigned to atoms unrelated to the bond were found in the partitioning line. If there was no maximum, this resulted in fractions very close to atom centers. This has been corrected. Another potential problem is if there is a maximum that is assigned to one of the atoms in the bond, but the maximum actually comes from a symmetrically equivalent but distinct atom. This would require significantly more work I think to fix so I'm leaving it for now.

SWeav02 committed:
Test fix for partitioning bug

SWeav02 committed:
black and isort

SWeav02 committed:
black and isort

SWeav02 committed:
Merge remote-tracking branch 'origin/basin_electride_finder' into basin_electride_finder

SWeav02 committed:
Extend basin assignment to within 0.3A of atom

Also added threads and check_for_covalency parameters.

SWeav02 committed:
Fix bug in voxel sphere finding method

SWeav02 committed:
black and isort

SWeav02 committed:
Fix typo with threads parameter

SWeav02 committed:
Fixed small bugs in electride finder and partitioning

rouding error in electride finder and small oversite in partitioning algorithm

SWeav02 committed:
Black and Isort

SWeav02 committed:
Fixed labeling bug in electride finder

The default label in ndimage doesn't consider corner connected voxels as connected. allowing this increases the speed of the algorithm heavily when there are many varied assignments from multi-site voxel assignment. Also, this indicates that there is an issue with the multi site assignment that needs fixing eventually. I think the problem stems from too many planes being considered when finding the closest plane.

SWeav02 committed:
Fix bug with multi-site assignment

An error was thrown if there were no voxels outside or on the partitioning planes.

SWeav02 committed:
Add savgol filter to linear step of partitioning

In some cases I find wavy patterns when using linear interpolation which messes with where minima and maxima are found. To deal with this, I put the results of the linear interpolation through a savgol filter.

SWeav02 committed:
fix dimensionality alg

SWeav02 committed:
Cleanup partitioning

SWeav02 committed:
reduce plane test

SWeav02 committed:
Reduce max multi-site atom distance

SWeav02 committed:
Add reduced partitioning

For the initial voxel assignment, the reduced planes aren't enough (at least in their current version). However, the large number of planes before reduction result in far too many planes to be able to determine which are relavent for the remaining voxel assignment. To deal with this I'm using the full partitioning for the first assignment and the reduced partitioning for the second part. It's possible this may cause issues if one of the removed planes is in fact important, but I think this is the best bet for now.

SWeav02 committed:
fix divide by 0 in reduced partitioning

SWeav02 committed:
fix by divide by 0 (correctly)

SWeav02 committed:
update multi-site atom distance

In some cases, the cutoff for the multi-site atom assignment was too small. This happened in the Hf2X type electrides. The radius to consider the X atom ends up being their elf ionic radii which results in a thin line connecting electride sites to make it 3D. Increasing the radius to 1.2*(furthest_neigh_dist-neigh_radii) fixes this issue.

SWeav02 committed:
Switched to using largest ELF radius for multi-site-assignment

SWeav02 committed:
Add method to get voxels at surface of sphere

SWeav02 committed:
Heavy Change to Unassigned Voxels

Assigning voxels to the nearest plane seems to be impossible or very difficult/computationally expensive as you would need to know the extent of the plane and calculate the distance from each voxel to the plane segment. Instead I've switched to a different method that relies on indexing rather than mathematical operations. The assignments of the voxels in a small sphere around the target voxels are found and the ratio of these assignments is found. If no nearby assignments are found, the sphere is made slightly larger and the operation repeats. This appears to give equal charges while also assigning the remaining voxels as I would expect.

SWeav02 committed:
Fixed crash when no voxels are outside partitioning

SWeav02 committed:
Black and Isort

SWeav02 committed:
Fix small bug in dimensionality alg

The structure used to label features in the grid was incorrect. It didn't include corner sharing voxels as being connected.

SWeav02 committed:
Fixed electride sites being placed at very small volumes

SWeav02 committed:
Updated grid maxima finder

SWeav02 committed:
Fix bug from pymatgen's merge_sites

SWeav02 committed:
Add pybader dependency

SWeav02 committed:
Merge pull request #4 from SWeav02/basin_electride_finder

Switch to pybader

jacksund commented:
> One thing to note is that I switched from the Henkelman group's code to pybader over the last couple months, which adds pybader as a dependency

sounds good. `pybader` will stay as an optional dependency until the bader app itself becomes a default. Also, it looks like pybader is already on conda-forge which is nice, but the package itself hasn't been updated in 4yrs. This might cause issues down the road. For example, if pybader can't run on python 3.12 or some other future version, the rest of the simmate repo will keep moving forward while the bader app gets left behind / disabled entirely. Their package is pretty small so creating a fork (or even absorbing it into your app module) is reasonable if it comes to that

> I'm hoping to merge this sometime soon

🥳

SWeav02 committed:
Update Docs

SWeav02 committed:
fix doc quotes

SWeav02 committed:
Merge branch 'main' into update_electride_search

SWeav02 committed:
Add minimum size for electride electrons

Sometimes a very small basin is registered as an electride when it is more likely just a dft or voxelation error. Adding a small cutoff of 0.1 A^3 should help with this.

SWeav02 committed:
Update badelf test name

SWeav02 commented:
@jacksund I just ran some tests for the BadELF workflows and they're running fine. I'm still unclear on the tests failing on black. Is the newest version of black still different from what we use in simmate?

Also, if possible I'd like to do a new release with this code so that Scott can use it. I can get started on running the full test suite.

jacksund commented:
no problem, I'll check out the black issue when I get the chance to review/merge this

btw this is unlucky timing because I will be traveling this weekend & am checking out in a bit. I'll try to get the new release out on Monday though

SWeav02 commented:
> btw this is unlucky timing because I will be traveling this weekend & am checking out in a bit. I'll try to get the new release out on Monday though

No problem, its not super urgent. Safe traveling!

jacksund committed:
run black

jacksund committed:
move optional deps, fix line endings, and format doc examples

jacksund committed:
add missing comma to pyproject

jacksund committed:
add comment to changelog

jacksund commented:
I made the following changes:
- pyrho and pybader are moved to optional dependencies. The docs for installing badelf were updated. There's also an error for when the deps aren't found but badelf is added to the registered apps
- files with windows line endings were changed back to unix (e.g. parameters.md)
- python examples for badelf were adjusted to have black formatting
- added a note to the changelog
